File: src/lib/services/preview.ts
/** Browser-only design preview. Database and import operations require Rust IPC. */
import type { Project, Position } from '$lib/types';
import sample from './sample.json';
const key = 'schema-atlas-browser-preview-v1';
function load(): Project[] {
  return JSON.parse(localStorage.getItem(key) ?? '[]') as Project[];
}
function save(projects: Project[]) {
  localStorage.setItem(key, JSON.stringify(projects));
}
export async function preview<T>(command: string, args: Record<string, unknown>): Promise<T> {
  const projects = load();
  let result: unknown;
  const project = projects.find((p) => p.id === args.projectId);
  switch (command) {
    case 'undo_canvas':
    case 'remove_canvas_group': {
      const source = project?.sources.find((s) => s.id === args.sourceId);
      if (!source || !project) throw new Error('Source not found.');
      if (command === 'undo_canvas') {
        if (!source.layoutBackup) throw new Error('No canvas edit to undo.');
        source.positions = source.layoutBackup.positions;
        source.groups = source.layoutBackup.groups;
        source.layoutBackup = null;
      } else {
        source.layoutBackup = {
          positions: structuredClone(source.positions),
          groups: structuredClone(source.groups ?? []),
        };
        source.groups = source.groups?.filter((g) => g.id !== args.groupId);
      }
      result = project;
      break;
    }

    case 'set_canvas_group': {
      const input = args.args as {
        source_id: string;
        group_id: string | null;
        name: string;
        node_ids: string[];
        color: string | null;
      };
      const source = project?.sources.find((s) => s.id === input.source_id);
      if (!source || !project) throw new Error('Source not found.');
      const previous = source.groups?.find((g) => g.id === input.group_id);
      if (input.group_id && !previous) throw new Error('Group not found.');
      const color = input.color ?? previous?.color ?? '#879b91';
      if (
        !input.name.trim() ||
        input.name.trim().length > 80 ||
        !/^#[0-9a-f]{6}$/i.test(color) ||
        !input.node_ids.length ||
        new Set(input.node_ids).size !== input.node_ids.length ||
        input.node_ids.some((id) => !source.graph.entities.some((e) => e.id === id))
      )
        throw new Error('Choose a name, color, and valid group members.');
      if (!previous && (source.groups?.length ?? 0) >= 100)
        throw new Error('Use at most 100 groups per source.');
      source.layoutBackup = {
        positions: structuredClone(source.positions),
        groups: structuredClone(source.groups ?? []),
      };
      const group = {
        id: previous?.id ?? crypto.randomUUID(),
        name: input.name.trim(),
        nodeIds: input.node_ids,
        color: color.toLowerCase(),
      };
      source.groups = previous
        ? source.groups!.map((g) => (g.id === group.id ? group : g))
        : [...(source.groups ?? []), group];
      result = project;
      break;
    }
    case 'list_projects':
      return projects as T;
    case 'create_demo': {
      const demo = structuredClone(sample) as Project;
      demo.id = crypto.randomUUID();
      demo.createdAt = demo.updatedAt = new Date().toISOString();
      for (const source of demo.sources) source.id = crypto.randomUUID();
      projects.push(demo);
      result = demo;
      break;
    }
    case 'create_project': {
      const p: Project = {
        id: crypto.randomUUID(),
        name: String(args.name).trim(),
        description: String(args.description),
        createdAt: new Date().toISOString(),
        updatedAt: new Date().toISOString(),
        sources: [],
      };
      projects.push(p);
      result = p;
      break;
    }
    case 'rename_project':
      if (project) {
        project.name = String(args.name);
        project.description = String(args.description);
        result = project;
      }
      break;
    case 'delete_project':
      save(projects.filter((p) => p.id !== args.projectId));
      return undefined as T;
    case 'remove_source':
      if (project) {
        project.sources = project.sources.filter((s) => s.id !== args.sourceId);
        result = project;
      }
      break;
    case 'save_positions': {
      const source = project?.sources.find((s) => s.id === args.sourceId);
      if (!source || !project) throw new Error('Source not found.');
      source.layoutBackup = {
        positions: structuredClone(source.positions),
        groups: structuredClone(source.groups ?? []),
      };
      source.positions = { ...source.positions, ...(args.positions as Record<string, Position>) };
      result = project;
      break;
    }
    default:
      throw new Error(
        'Open the Schematlas desktop app to connect databases or import OpenAPI files.',
      );
  }
  save(projects);
  return result as T;
}

File: ui/src/preview.rs
//! Browser-only design preview backed by localStorage. Database and import operations need the desktop runtime.
use crate::api::{local_storage_get, local_storage_set, now_iso};
use crate::types::{CanvasGroup, LayoutBackup, Position, Project, Source};
use serde_json::Value;
use std::collections::HashMap;

const KEY: &str = "schema-atlas-browser-preview-v1";
const SAMPLE: &str = include_str!("sample.json");

fn load() -> Vec<Project> {
    local_storage_get(KEY)
        .and_then(|raw| serde_json::from_str(&raw).ok())
        .unwrap_or_default()
}

fn save(projects: &[Project]) {
    if let Ok(raw) = serde_json::to_string(projects) {
        local_storage_set(KEY, &raw);
    }
}

fn text(value: &Value, key: &str) -> String {
    value
        .get(key)
        .and_then(Value::as_str)
        .unwrap_or_default()
        .to_string()
}

fn backup(source: &Source) -> LayoutBackup {
    LayoutBackup {
        positions: source.positions.clone(),
        groups: source.groups().to_vec(),
    }
}

fn to_value<T: serde::Serialize>(value: &T) -> Result<Value, String> {
    serde_json::to_value(value).map_err(|e| e.to_string())
}

pub fn preview(command: &str, args: Value) -> Result<Value, String> {
    let mut projects = load();
    let project_id = text(&args, "projectId");
    let index = projects.iter().position(|p| p.id == project_id);
    let result = match command {
        "list_projects" => return to_value(&projects),
        "create_demo" => {
            let mut demo: Project = serde_json::from_str(SAMPLE).map_err(|e| e.to_string())?;
            demo.id = uuid::Uuid::new_v4().to_string();
            demo.created_at = now_iso();
            demo.updated_at = demo.created_at.clone();
            for source in &mut demo.sources {
                source.id = uuid::Uuid::new_v4().to_string();
            }
            projects.push(demo.clone());
            to_value(&demo)?
        }
        "create_project" => {
            let project = Project {
                id: uuid::Uuid::new_v4().to_string(),
                name: text(&args, "name").trim().to_string(),
                description: text(&args, "description"),
                created_at: now_iso(),
                updated_at: now_iso(),
                sources: vec![],
                working_directory: None,
            };
            projects.push(project.clone());
            to_value(&project)?
        }
        "rename_project" => {
            let project = index
                .map(|i| &mut projects[i])
                .ok_or("Project not found.")?;
            project.name = text(&args, "name");
            project.description = text(&args, "description");
            to_value(project)?
        }
        "delete_project" => {
            projects.retain(|p| p.id != project_id);
            save(&projects);
            return Ok(Value::Null);
        }
        "remove_source" => {
            let source_id = text(&args, "sourceId");
            let project = index
                .map(|i| &mut projects[i])
                .ok_or("Project not found.")?;
            project.sources.retain(|s| s.id != source_id);
            to_value(project)?
        }
        "save_positions" => {
            let source_id = text(&args, "sourceId");
            let positions: HashMap<String, Position> =
                serde_json::from_value(args.get("positions").cloned().unwrap_or(Value::Null))
                    .map_err(|e| e.to_string())?;
            let project = index.map(|i| &mut projects[i]).ok_or("Source not found.")?;
            let source = project
                .sources
                .iter_mut()
                .find(|s| s.id == source_id)
                .ok_or("Source not found.")?;
            source.layout_backup = Some(backup(source));
            source.positions.extend(positions);
            to_value(project)?
        }
        "undo_canvas" | "remove_canvas_group" => {
            let source_id = text(&args, "sourceId");
            let project = index.map(|i| &mut projects[i]).ok_or("Source not found.")?;
            let source = project
                .sources
                .iter_mut()
                .find(|s| s.id == source_id)
                .ok_or("Source not found.")?;
            if command == "undo_canvas" {
                let restored = source
                    .layout_backup
                    .take()
                    .ok_or("No canvas edit to undo.")?;
                source.positions = restored.positions;
                source.groups = Some(restored.groups);
            } else {
                let group_id = text(&args, "groupId");
                source.layout_backup = Some(backup(source));
                source.groups = Some(
                    source
                        .groups()
                        .iter()
                        .filter(|g| g.id != group_id)
                        .cloned()
                        .collect(),
                );
            }
            to_value(project)?
        }
        "set_canvas_group" => {
            let input = args.get("args").cloned().unwrap_or(Value::Null);
            let source_id = text(&input, "source_id");
            let group_id = input
                .get("group_id")
                .and_then(Value::as_str)
                .map(str::to_string);
            let name = text(&input, "name").trim().to_string();
            let node_ids: Vec<String> = input
                .get("node_ids")
                .and_then(Value::as_array)
                .map(|a| {
                    a.iter()
                        .filter_map(Value::as_str)
                        .map(str::to_string)
                        .collect()
                })
                .unwrap_or_default();
            let project = index.map(|i| &mut projects[i]).ok_or("Source not found.")?;
            let source = project
                .sources
                .iter_mut()
                .find(|s| s.id == source_id)
                .ok_or("Source not found.")?;
            let previous = source
                .groups()
                .iter()
                .find(|g| Some(&g.id) == group_id.as_ref())
                .cloned();
            if group_id.is_some() && previous.is_none() {
                return Err("Group not found.".into());
            }
            let color = input
                .get("color")
                .and_then(Value::as_str)
                .map(str::to_string)
                .or_else(|| previous.as_ref().and_then(|g| g.color.clone()))
                .unwrap_or_else(|| crate::layout::DEFAULT_GROUP_COLOR.to_string());
            let mut unique = node_ids.clone();
            unique.sort();
            unique.dedup();
            let valid_color = color.len() == 7
                && color.starts_with('#')
                && color[1..].chars().all(|c| c.is_ascii_hexdigit());
            if name.is_empty()
                || name.len() > 80
                || !valid_color
                || node_ids.is_empty()
                || unique.len() != node_ids.len()
                || node_ids
                    .iter()
                    .any(|id| !source.graph.entities.iter().any(|e| &e.id == id))
            {
                return Err("Choose a name, color, and valid group members.".into());
            }
            if previous.is_none() && source.groups().len() >= 100 {
                return Err("Use at most 100 groups per source.".into());
            }
            source.layout_backup = Some(backup(source));
            let group = CanvasGroup {
                id: previous
                    .as_ref()
                    .map(|g| g.id.clone())
                    .unwrap_or_else(|| uuid::Uuid::new_v4().to_string()),
                name,
                node_ids,
                color: Some(color.to_lowercase()),
            };
            let mut groups = source.groups().to_vec();
            if previous.is_some() {
                for g in &mut groups {
                    if g.id == group.id {
                        *g = group.clone();
                    }
                }
            } else {
                groups.push(group);
            }
            source.groups = Some(groups);
            to_value(project)?
        }
        _ => {
            return Err(
                "Open the Schematlas desktop app to connect databases or import OpenAPI files."
                    .into(),
            )
        }
    };
    save(&projects);
    Ok(result)
}
